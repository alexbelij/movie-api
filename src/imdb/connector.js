/* @flow */

import DataLoader from 'dataloader';
import PromiseThrottle from 'promise-throttle';
import rp from 'request-promise-native';

import { userAgent } from '../utils';
import env from '../env';
import HtmlConnector from '../HtmlConnector';

const IMDB_ROOT = 'http://www.imdb.com';

export type ImdbConnectorConfig = {
  userId?: string,
};

type Rp = (options: Object) => Promise<any>;

class ImdbConnector extends HtmlConnector {
  _userId: string;
  _ratingsRp: Rp;

  constructor({ userId }: ImdbConnectorConfig = {}) {
    super({
      rootUrl: IMDB_ROOT,
      rps: 2,
    });

    this._userId = userId || env.getImdbUserId();

    this._ratingsRp = rp.defaults({
      headers: { 'User-Agent': userAgent },
      gzip: true,
      qs: {
        u: this._userId,
      },
    });
  }

  _ratingsThrottleQueue = new PromiseThrottle({
    requestsPerSecond: 10,
    promiseImplementation: Promise,
  });

  ratingsLoader: { load: (imdbId: string) => Promise<any> } = new DataLoader(
    (imdbIds: Array<string>) => this._ratingsThrottleQueue.addAll(
      imdbIds.map((imdbId: string) => () => this._ratingsRp({
        // eslint-disable-next-line max-len
        uri: `http://p.media-imdb.com/static-content/documents/v1/title/${imdbId}/ratings%3Fjsonp=imdb.rating.run:imdb.api.title.ratings/data.json`,
      })),
    ), {
      batch: false,
    },
  );

  ratingsGet = (imdbId: string) => this.ratingsLoader.load(imdbId);
}

export default ImdbConnector;
