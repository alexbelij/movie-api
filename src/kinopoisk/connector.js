/* @flow */

import crypto from 'crypto';
import querystring from 'querystring';

import DataLoader from 'dataloader';
import moment from 'moment';
import R from 'ramda';
import rp from 'request-promise-native';

import { userAgent } from '../utils';

const KINOPOISK_API_ROOT = 'https://ext.kinopoisk.ru/ios/3.11.0';
const KINOPOISK_API_KEY = 'a17qbcw1du0aedm';
const KINOPOISK_PLUS_ROOT = 'https://plus.kinopoisk.ru';

type Endpoint =
 // Movie info
 | 'getGallery'
 | 'getStaffList'
 | 'getKPFilmDetailView'
 | 'getKPFilmsList'
 | 'getKPPeopleDetailView'
 // Genres
 | 'getKPGenresView'
 | 'getKPTopGenre'
 // Reviews
 | 'getKPReviews'
 | 'getKPReviewDetail'
 // News
 | 'getKPNewsView'
 | 'getKPNewsDetail'
 // Coming soon
 | 'getKPSoonFilms'
 | 'getKPSoonDVD'
 | 'getDatesForSoonFilms'
 | 'getDatesForSoonDVD'
 // Showtimes
 | 'getKPTodayFilms'
 | 'getKPCinemas'
 | 'getKPCinemaDetailView'
 | 'getSeance'
 | 'getDatesForDetailCinema'
 | 'getDatesForSeance'
 // Top
 | 'getKPTop'
 | 'getPopularIndex'
 // Search
 | 'getKPGlobalSearch'
 | 'getKPLiveSearch'
 | 'getKPSearchInCinemas'
 | 'getKPSearchInFilms'
 | 'getKPSearchInPeople'
 | 'navigator'
 | 'navigatorFilters'
 // Geo Reference
 | 'getKPCountryView'
 | 'getKPAllCitiesView';

const applyQueryToUrl = (
  url: string,
  query: { [key: string]: mixed } = {},
) => `${url}?${querystring.stringify(query)}`;

type Rp = (options: Object) => Promise<any>;
type Loader = { load: (url: string) => Promise<any> };

class KinopoiskConnector {
  _apiRp: Rp = rp.defaults({
    headers: {
      'Android-Api-Version': '22',
      'Cache-Control': 'max-stale=0',
      'Image-Scale': '3',
      'User-Agent': 'Android client (5.1 / api22), ru.kinopoisk/3.7.0 (45)',
      cityID: '1',
      clientDate: moment(new Date()).format('HH:mm MM.dd.YYYY'),
      ClientId: crypto
        .createHash('md5')
        .update(String(Math.floor(Math.random() * (99999 + 1))))
        .digest('hex'),
      countryID: '2',
      device: 'android',
    },
    timeout: 4000,
    json: true,
    gzip: true,
    jar: true,
  });

  _htmlRp: Rp = rp.defaults({
    headers: { 'User-Agent': userAgent },
    gzip: true,
  });

  apiLoader: Loader = new DataLoader(
    (urls: Array<string>) => Promise.all(
      urls.map(async (url: string) => this._apiRp({
        uri: url,
        qs: {
          key: crypto
            .createHash('md5')
            // eslint-disable-next-line max-len
            .update(`${url.replace(`${KINOPOISK_API_ROOT}/`, '')}${KINOPOISK_API_KEY}`)
            .digest('hex'),
        },
      })),
    ), {
      batch: false,
    },
  );

  htmlLoader: Loader = new DataLoader(
    (urls: Array<string>) => Promise.all(
      urls.map((url: string) => this._htmlRp({ uri: url })),
    ), {
      batch: false,
    },
  );

  apiGet = async (
    endpoint: Endpoint,
    query: void | { [key: string]: mixed },
  ) => {
    const json = await this.apiLoader.load(
      applyQueryToUrl(
        `${KINOPOISK_API_ROOT}/${endpoint}`, query,
      ),
    );

    return R.prop('data')(json);
  };

  htmlGet = (
    path: string,
    query: void | { [key: string]: mixed },
  ) => this.htmlLoader.load(
    applyQueryToUrl(`${KINOPOISK_PLUS_ROOT}/${path}`, query),
  );
}

export default KinopoiskConnector;
