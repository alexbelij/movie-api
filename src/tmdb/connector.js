/* @flow */

import DataLoader from 'dataloader';
import PromiseThrottle from 'promise-throttle';
import rp from 'request-promise-native';

import {userAgent} from '../utils';
import env from '../env';

const TMDB_API_ROOT = 'https://api.themoviedb.org/3';

export type TmdbConnectorConfig = {
  apiKey?: string,
  language: string,
};

class TmdbConnector {
  _apiKey: string;
  _language: string;
  _rp: (options: Object) => Promise<any>;

  constructor({apiKey, language = 'ru'}: TmdbConnectorConfig = {}) {
    this._apiKey = apiKey || env.getTmdbApiKey();
    this._language = language;

    this._rp = rp.defaults({
      headers: {'User-Agent': userAgent},
      gzip: true,
      qs: {
        api_key: this._apiKey,
        language: this._language,
      },
      json: true,
    });
  }

  _throttleQueue = new PromiseThrottle({
    requestsPerSecond: 3, // TMDB's rate limit is 40 requests / 10 seconds
    promiseImplementation: Promise,
  });

  apiLoader: DataLoader<string, any> = new DataLoader(
    (optionsHashes: Array<string>) =>
      this._throttleQueue.addAll(
        optionsHashes.map((optionsHash: string) => () =>
          this._rp(JSON.parse(optionsHash)),
        ),
      ),
    {
      batch: false,
    },
  );

  apiGet = (endpoint: string, query: {[key: string]: mixed}) =>
    this.apiLoader.load(
      JSON.stringify({
        uri: `${TMDB_API_ROOT}/${endpoint}`,
        qs: query,
      }),
    );
}

export default TmdbConnector;
