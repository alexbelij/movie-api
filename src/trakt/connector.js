/* @flow */

import DataLoader from 'dataloader';
import PromiseThrottle from 'promise-throttle';
import rp from 'request-promise-native';

import {userAgent} from '../utils';
import env from '../env';

const TRAKT_API_ROOT = 'https://api.trakt.tv';
const TRAKT_API_VERSION = 2;

export type TraktConnectorConfig = {
  apiKey?: string,
};

class TraktConnector {
  _apiKey: string;
  _language: string;
  _rp: (options: Object) => Promise<any>;

  constructor({apiKey}: TraktConnectorConfig = {}) {
    this._apiKey = apiKey || env.getTraktApiKey();

    this._rp = rp.defaults({
      headers: {
        'User-Agent': userAgent,
        'trakt-api-version': TRAKT_API_VERSION,
        'trakt-api-key': this._apiKey,
      },
      gzip: true,
      json: true,
    });
  }

  _throttleQueue = new PromiseThrottle({
    requestsPerSecond: 3,
    promiseImplementation: Promise,
  });

  apiLoader: DataLoader<string, any> = new DataLoader(
    (optionsHashes: Array<string>) =>
      this._throttleQueue.addAll(
        optionsHashes.map((optionsHash: string) => () =>
          this._rp(JSON.parse(optionsHash)),
        ),
      ),
    {
      batch: false,
    },
  );

  apiGet = (endpoint: string, query: void | {[key: string]: mixed}) =>
    this.apiLoader.load(
      JSON.stringify({
        uri: `${TRAKT_API_ROOT}/${endpoint}`,
        qs: query || {},
      }),
    );
}

export default TraktConnector;
