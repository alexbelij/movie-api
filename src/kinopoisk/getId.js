/* @flow */

import cheerio from 'cheerio';
import R from 'ramda';
import similarity from 'similarity';

import connector from './connector';

type Query = {
  title: string,
  year?: number,
  countries?: Array<string>,
  isTvShow?: boolean,
};
type SearchResult = {
  id: number,
  title: string,
  countries: Array<string>,
  year: number,
};

const scrapeResults = (
  query: Query,
  html: string,
): Array<SearchResult> => {
  const $ = cheerio.load(html);

  return $(`.film-snippet_type_${query.isTvShow ? 'show' : 'movie'}`)
    .get()
    .map((snippet: any) => {
      const titleNode = $(snippet).find('.film-snippet__title-link');
      const id = parseInt(
        R.nth(1, titleNode.attr('href').match(/\/film\/(.+)\//)),
        10,
      );

      if (isNaN(id)) return null;

      const infoNode = $(snippet).find('.film-snippet__info');
      const title = titleNode.text().replace(/\s\s+/g, ' ');
      const countries = R.pipe(
        R.split(','),
        R.map(R.trim),
        R.filter(
          (country: ?string) => !!country && isNaN(parseInt(country, 10)),
        ),
      )(infoNode.text());
      const year = parseInt(R.head(infoNode.text().match(/\d+/)), 10);

      return { id, title, countries, year };
    });
};

const filterResults = (
  query: Query,
  results: Array<?SearchResult>,
): Array<SearchResult> => R.filter(R.allPass([
  Boolean,
  ({ title }: SearchResult) => similarity(title, query.title) >= 0.8,
  (query.year
    ? ({ year }: SearchResult) => year === query.year
    : R.always(true)),
  (query.countries
    ? ({ countries }: SearchResult) => R.equals(
        R.intersection(countries, query.countries).length,
        // $FlowFixMe: countries is always available here
        query.countries.length,
      )
    : R.always(true)),
]))(results);

const getId = async (query: Query): Promise<?number> => {
  const html = await connector.htmlGet('search', { text: query.title });

  return R.pipe(
    R.curry(scrapeResults)(query),
    R.curry(filterResults)(query),
    R.sortBy(({ title }: SearchResult) => similarity(title, query.title)),
    R.head,
    R.prop('id'),
  )(html);
};

export {
  scrapeResults as __scrapeResults,
  filterResults as __filterResults,
};
export default getId;
