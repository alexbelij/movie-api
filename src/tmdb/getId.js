/* @flow */

import R from 'ramda';

import connector from './connector';

type Query = {
  title?: string,
  year?: number,
  imdbId?: string,
  isTvShow?: boolean,
};

const findByImdbId = async (query: Query): Promise<?number> => {
  if (!query.imdbId) return null;

  const res = await connector.apiGet(
    `find/${query.imdbId}`,
    { external_source: 'imdb_id' },
  );

  const results = query.isTvShow
    ? R.propOr([], 'tv_results', res)
    : R.propOr([], 'movie_results', res);

  return R.pipe(R.head, R.prop('id'))(results);
};
const searchByTitle = async (query: Query): Promise<?number> => {
  if (!query.title) return null;

  const res = await connector.apiGet(
    `search/${query.isTvShow ? 'tv' : 'movie'}`,
    { query: query.title, year: query.year },
  );

  return R.pipe(R.prop('results'), R.head, R.prop('id'))(res);
};

const getId = async (query: Query): Promise<?number> => {
  if (query.imdbId) return findByImdbId(query);
  if (query.title) return searchByTitle(query);

  return null;
};

export {
  findByImdbId as __findByImdbId,
  searchByTitle as __searchByTitle,
};
export default getId;
