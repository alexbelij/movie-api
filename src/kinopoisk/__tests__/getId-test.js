/* @flow */

import getId, {
  __scrapeResults,
  __filterResults,
} from '../getId';
import connector from '../connector';

const movieQuery = {
  title: 'Звёздные войны: Пробуждение силы',
};
const tvShowQuery = {
  title: 'Игра престолов',
  isTvShow: true,
};

describe('getId', () => {
  it('scrapes movie search results', async () => {
    expect(__scrapeResults(
      movieQuery,
      await connector.htmlGet('search', { text: movieQuery.title }),
    )).toMatchSnapshot();
  });

  it('scrapes tv show search results', async () => {
    expect(__scrapeResults(
      tvShowQuery,
      await connector.htmlGet('search', { text: tvShowQuery.title }),
    )).toMatchSnapshot();
  });

  it('filters results', () => {
    expect(__filterResults({ title: 'foo' }, [
      null,
      undefined,
      { id: 1, title: 'foo', countries: [], year: 2016 },
    ])).toMatchSnapshot();

    expect(__filterResults({ title: 'Звёздные войны' }, [
      { id: 1, title: 'Звёздные войны', countries: [], year: 2015 },
      { id: 2, title: 'Звездные войны', countries: [], year: 2016 },
      { id: 3, title: 'Звёздная пыль', countries: [], year: 2014 },
      { id: 4, title: 'Звёздный путь', countries: [], year: 2016 },
    ])).toMatchSnapshot();

    expect(__filterResults({ title: 'foo', year: 2016 }, [
      { id: 1, title: 'foo', countries: [], year: 2015 },
      { id: 2, title: 'foo', countries: [], year: 2016 },
      { id: 3, title: 'foo', countries: [], year: 2014 },
      { id: 4, title: 'foo', countries: [], year: 2016 },
    ])).toMatchSnapshot();

    expect(__filterResults({ title: 'foo', countries: ['США', 'Россия'] }, [
      { id: 1, title: 'foo', countries: ['США'], year: 2016 },
      // eslint-disable-next-line max-len
      { id: 2, title: 'foo', countries: ['США', 'Россия', 'Германия'], year: 2016 },
      { id: 3, title: 'foo', countries: ['Россия'], year: 2016 },
      { id: 4, title: 'foo', countries: ['США', 'Россия'], year: 2016 },
    ])).toMatchSnapshot();
  });

  it('finds best movie id for a given query', async () => {
    expect(await getId(movieQuery)).toMatchSnapshot();
  });

  it('finds best tv show id for a given query', async () => {
    expect(await getId(tvShowQuery)).toMatchSnapshot();
  });
});
