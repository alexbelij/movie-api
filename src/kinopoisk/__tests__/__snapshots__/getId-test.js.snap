exports[`getId filters results 1`] = `
Array [
  Object {
    "countries": Array [],
    "id": 1,
    "title": "foo",
    "year": 2016,
  },
]
`;

exports[`getId filters results 2`] = `
Array [
  Object {
    "countries": Array [],
    "id": 1,
    "title": "Звёздные войны",
    "year": 2015,
  },
  Object {
    "countries": Array [],
    "id": 2,
    "title": "Звездные войны",
    "year": 2016,
  },
]
`;

exports[`getId filters results 3`] = `
Array [
  Object {
    "countries": Array [],
    "id": 2,
    "title": "foo",
    "year": 2016,
  },
  Object {
    "countries": Array [],
    "id": 4,
    "title": "foo",
    "year": 2016,
  },
]
`;

exports[`getId filters results 4`] = `
Array [
  Object {
    "countries": Array [
      "США",
      "Россия",
      "Германия",
    ],
    "id": 2,
    "title": "foo",
    "year": 2016,
  },
  Object {
    "countries": Array [
      "США",
      "Россия",
    ],
    "id": 4,
    "title": "foo",
    "year": 2016,
  },
]
`;

exports[`getId finds best movie id for a given query 1`] = `714888`;

exports[`getId finds best tv show id for a given query 1`] = `464963`;

exports[`getId scrapes movie search results 1`] = `
Array [
  Object {
    "countries": Array [
      "США",
    ],
    "id": 714888,
    "title": "Звёздные войны: Пробуждение силы",
    "year": 2015,
  },
]
`;

exports[`getId scrapes tv show search results 1`] = `
Array [
  Object {
    "countries": Array [
      "США",
      "Великобритания",
    ],
    "id": 464963,
    "title": "Игра престолов",
    "year": 2011,
  },
]
`;
